File: bigbear/615/186.java
class Solution {
    public void reverseWords(char[] str) {
        reverse(str, 0, str.length - 1);
        int start = 0;
        for(int i = 0; i < str.length; i++){
            if(str[i] == ' '){
                reverse(str, start, i - 1);
                start = i + 1;
            }
        }
        reverse(str, start, str.length - 1);
    }
    
    private void reverse(char[] str, int start, int end){
        while(start < end){
            char tmp = str[start];
            str[start] = str[end];
            str[end] = tmp;
            start++;
            end--;
        }
    }
}

File: zemiao/186.py
class Solution:
    """
    @param str: a string
    @return: return a string
    """
    def reverseWords(self, str):
      begin, end = 0,len(str)
      l,r = begin,end-1
      while l<end and str[l]!=' ':
        l += 1
      while r>=0 and str[r]!=' ':
        r -= 1
      if l>r:
        return str
      if l==r:
        return str[r+1:end]+' '+str[begin:l]
      return str[r+1:end]+' '+self.reverseWords(str[l+1:r])+' '+str[begin:l]
